pkg/services/ngalert/remote/client/mimir.go

```diff
@@ -0,0 +1,184 @@
+	SetGrafanaAlertmanagerConfiguration()
+	DeleteGrafanaAlertmanagerConfiguration()
+
+	GetGrafanaAlertmanagerState()
+	SetGrafanaAlertmanagerState()
+	DeleteGrafanaAlertmanagerState()
+}
+
+type Mimir struct {
+	endpoint *url.URL
+	client   http.Client
+	logger   log.Logger
+}
+
+type Config struct {
+	Address  string
+	TenantID string
+	Password string
+
+	Logger log.Logger
+}
+
+// successResponse represents a successful response from the Mimir API.
+type successResponse struct {
+	Status string `json:"status"`
+	Data   any    `json:"data"`
+}
+
+// errorResponse represents an error from the Mimir API.
+type errorResponse struct {
+	Status string `json:"status"`
+	Error1 string `json:"error"`
+	Error2 string `json:"Error"`
+}
+
+func (e *errorResponse) Error() string {
+	if e.Error1 != "" {
+		return e.Error1
+	}
+
+	return e.Error2
+}
+
+func New(cfg *Config) (*Mimir, error) {
+	endpoint, err := url.Parse(cfg.Address)
+	if err != nil {
+		return nil, err
+	}
+
+	rt := &MimirAuthRoundTripper{
+		TenantID: cfg.TenantID,
+		Password: cfg.Password,
+		Next:     http.DefaultTransport,
+	}
+
+	c := http.Client{
+		Transport: rt,
+	}
+
+	return &Mimir{
+		endpoint: endpoint,
+		client:   c,
+		logger:   cfg.Logger,
+	}, nil
+}
+
+// do execute an HTTP requests against the specified path and method using the specified payload. It returns the HTTP response.
+func (mc *Mimir) do(ctx context.Context, p, method string, payload io.Reader, contentLength int64, out any) (*http.Response, error) {
+	pathURL, err := url.Parse(p)
+	if err != nil {
+		return nil, err
+	}
+
+	endpoint := *mc.endpoint
+	endpoint.Path = path.Join(endpoint.Path, pathURL.Path)
+
+	r, err := http.NewRequestWithContext(ctx, method, endpoint.String(), payload)
+	if err != nil {
+		return nil, err
+	}
+
+	r.Header.Set("Accept", "application/json")
+	r.Header.Set("Content-Type", "application/json")
+
+	if contentLength > 0 {
+		r.ContentLength = contentLength
+	}
+
+	logger := mc.logger.New("url", r.URL.String(), "method", r.Method)
+	resp, err := mc.client.Do(r)
+	if err != nil {
+		msg := "unable to fulfill request to the Mimir API"
+		logger.Error(msg, "err", err)
+		return nil, fmt.Errorf("%s: %w", msg, err)
+	}
+	defer func() {
+		if err := resp.Body.Close(); err != nil {
+			logger.Error("error closing HTTP body", "err", err)
+		}
+	}()
+
+	logger = logger.New("status", resp.StatusCode)
+	ct := resp.Header.Get("Content-Type")
+	if !strings.HasPrefix(ct, "application/json") {
+		msg := "response content-type is not application/json"
+		logger.Error(msg, "content-type", ct)
+		return nil, fmt.Errorf("%s: %s", msg, ct)
+	}
+
+	if out == nil {
+		return resp, nil
+	}
+
+	body, err := io.ReadAll(resp.Body)
+	if err != nil {
+		msg := "failed to read the request body"
+		logger.Error(msg, "err", err)
+		return nil, fmt.Errorf("%s: %w", msg, err)
+	}
+
+	if resp.StatusCode/100 != 2 {
+		errResponse := &errorResponse{}
+		jsonErr := json.Unmarshal(body, errResponse)
```

Comment:
Just use `err`, the name `jsonErr` implies that this is a different type.
```suggestion
		err := json.Unmarshal(body, errResponse)
```